src/scaler/mod.rs

```diff
@@ -18,15 +20,37 @@ pub mod spreadscaler;
 /// to determine if actions need to be taken in response to an event
 #[async_trait]
 pub trait Scaler {
-    type Config: Send + Sync;
-
-    /// Provide a scaler with configuration to use internally when computing commands
-    /// This should trigger a reconcile with the new configuration
-    async fn update_config(&mut self, config: Self::Config) -> Result<Vec<Command>>;
+    /// Provide a scaler with configuration to use internally when computing commands This should
+    /// trigger a reconcile with the new configuration.
+    ///
+    /// This config can be anything that can be turned into a
+    /// [`TraitProperty`](crate::model::TraitProperty). Additional configuration outside of what is
+    /// available in a `TraitProperty` can be passed when constructing the scaler
+    async fn update_config(&mut self, config: TraitProperty) -> Result<Vec<Command>>;
 
     /// Compute commands that must be taken given an event that changes the lattice state
     async fn handle_event(&self, event: &Event) -> Result<Vec<Command>>;
 
     /// Compute commands that must be taken to achieve desired state as specified in config
     async fn reconcile(&self) -> Result<Vec<Command>>;
+
+    /// Returns the list of commands needed to cleanup for a scaler
+    ///
+    /// This purposefully does not consume the scaler so that if there is a failure it can be kept
+    /// around
+    async fn cleanup(&self) -> Result<Vec<Command>>;
+
+    /// An optional to implement function that can tell the scaler to enter a backoff period. When
+    /// in backoff, a scaler should always return an empty `Vec<Command>` for all operations. Please
+    /// be careful when calling backoff until you have sent the commands to their intended location
+    /// as otherwise the might not rerun until they are reconciled again.
+    ///
+    /// This function should return as quickly as possible, spawning a task that can switch the
+    /// state back from backoff. The passed in channel is meant to be used to notify the caller that
+    /// the backoff is complete with the name of the model. This name can then be used by the caller
+    /// to run a reconciliation pass to ensure everything is done. This is intended to workaround
+    /// the fact that there is no way to tell when we finished stopping/starting actors
+    ///
+    /// Please note that we'd like this to be temporary in the long term
```

Comment:
I think this will be the hardest thing for other folks to do _correctly_ if they end up creating scalers to extend wadm, and I definitely like the idea of us dropping this. Putting this here so I remember to put this in #86